src/pip/_internal/req/req_file.py

```diff
@@ -537,6 +537,13 @@ def get_file_content(url: str, session: PipSession) -> Tuple[str, str]:
 
     # Assume this is a bare path.
     try:
+        if not os.path.exists(url):
+            # If the expected bare path doesn't exist, check if treating symlink'd path
+            # components as real works.
+            abs_path_url = os.path.abspath(url)
+            if os.path.exists(abs_path_url):
+                url = abs_path_url
```

Comment:
I'm not comfortable with the heuristics that triggers `if not os.path.exists`. Could this lead to different behaviors depending on, say, where the project has been checked out? Although I have not thought it through, this raises security warnings in my head.

Perhaps a better place to do handle this case is in `_parse_and_recurse` around where we do the `os.path.join`.

However I am not even sure of what the correct behaviour should be.
Actually in `_parse_and_recurse` we do a `urllib.parse.urljoin` if the base requirement comes from a URL.
Perhaps we should investigate what urljoin does in presence of `file://` URLs and symlinks and align the behavior with local paths?

Reply:
Cross referencing https://github.com/pypa/pip/issues/8765 which also discusses quircks in pip's handling of nested requirement files.

Reply:
@sbidoul I was looking at that location originally. I thought, however, that this was more specific, since path handling where the the path is a relative path AND the relative path might have multiple interpretations is specific to a bare file path, not a URL, even a file URL I think. Does `urljoin` actually examine any resources, or is it just string manipulation? If the latter, can it anticipate paths that might have multiple interpretations?

I am also trying to anticipate the security implications. I think a security issue would have to somehow find a way to create the condition of `os.path.exists(url) is False` and `os.path.exists(os.path.abspath(url))` is True` and `open(url)` would be successful. I don't see how that can be achieved.

I have also tried to solve our current issue using environment variables. That hasn't been successful because we have encountered cloud environments where the cloud API does now allow adding environment variables to the build phase of the container deployment process.

Reply:
I looked again at the structure that your test produces. It looks like this:

```
package1/requirements.txt
package2/requirements.txt  # has -r ../../package1/requirements.txt
...
```

So when we look at package2/requirements.txt, it contains an *invalid* relative reference to package1.

The test then goes on with creating a symlink structure that makes that relative reference look correct *if we don't resolve* some symlinks? Do I read it correctly?

As I understand it at this point, I'm really not comfortable with getting this change in. The resolution of nested requirement files should not lead to different results depending on the presence of absence of specific files.

Reply:
I left a comment in the issue:
https://github.com/pypa/pip/issues/11668#issuecomment-1407962507